components/support/jwcrypto/src/aes.rs

```diff
@@ -0,0 +1,84 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// Wrappers we use around rc_crypto's AES implementation. Specifically,
+// "enc=A256GCM" from RFC7518, Section 4.7 - for all the gory details, see
+// https://tools.ietf.org/html/rfc7518#section-4.7.
+
+use crate::{
+    error::{JwCryptoError, Result},
+    CompactJwe, EncryptionAlgorithm, JweHeader,
+};
+use rc_crypto::{aead, rand};
+
+/// Does the AES-encrypt heavy-lifting for the schemes supported by this crate.
+pub(crate) fn aes_gcm_encrypt(
+    data: &[u8],
+    protected_header: JweHeader,
+    secret: &[u8],
+) -> Result<CompactJwe> {
+    let encryption_algorithm = match protected_header.enc {
+        EncryptionAlgorithm::A256GCM => &aead::AES_256_GCM,
+    };
+    let sealing_key = aead::SealingKey::new(encryption_algorithm, secret)?;
+    let additional_data = serde_json::to_string(&protected_header)?;
+    let additional_data =
+        base64::encode_config(additional_data.as_bytes(), base64::URL_SAFE_NO_PAD);
+    let additional_data = additional_data.as_bytes();
+    let aad = aead::Aad::from(additional_data);
+    // Note that RFC7518 specifies an IV of 96 bits == 12 bytes - which means
+    // that a random IV generally isn't safe with AESGCM due to the risk of
+    // collisions in this many bits. However, for the use-cases supported by
+    // this crate, the keys are either ephemeral, or the number of encryptions
+    // for the same key is expected to be low enough to not collide in
+    // practice.
+    let mut iv: Vec<u8> = vec![0; 12];
+    rand::fill(&mut iv)?;
+    let nonce = aead::Nonce::try_assume_unique_for_key(encryption_algorithm, &iv)?;
+    let mut encrypted = aead::seal(&sealing_key, nonce, aad, data)?;
+
+    let tag_idx = encrypted.len() - encryption_algorithm.tag_len();
+    let auth_tag = encrypted.split_off(tag_idx);
+    let ciphertext = encrypted;
+
+    CompactJwe::new(
+        Some(protected_header),
+        None,
+        Some(iv),
+        ciphertext,
+        Some(auth_tag),
+    )
+}
+
+/// Does the AES-decrypt heavy-lifting for the schemes supported by this crate
+pub(crate) fn aes_gcm_decrypt(jwe: &CompactJwe, secret: &[u8]) -> Result<String> {
+    if jwe.encrypted_key()?.is_some() {
+        return Err(JwCryptoError::IllegalState(
+            "The Encrypted Key must be empty.",
+        ));
+    }
```

Comment:
nit: technically, it's not the `enc` algorithm's job to check that this is empty. Whether or not we expect a value here depends on the `alg` parameter. It is important that we do this check *somewhere*, so we could consider moving it up into the two `alg=ECDH-ES` and `alg=dir` implementations if there's a good place for it, or I'm happy to keep it here for simplicity and add a comment saying "not really the enc algorithm's job to check this but it's here for simplicity".